designs/disruption-controls-by-method.md

```diff
@@ -0,0 +1,397 @@
+* 👍 Provides place per action for generic controls intended to be shared across all disruption methods. While not all methods share the same disruption actions, there are already cases for this. 
+* 👍 Could extend other fields beyond generic values for specific actions with validation  
+* 👍 Allows for very natural reason specification for further granularity of specifying specific K8sVersion upgrades for example. There is no need for complicated validation in regards to what method types are compatible with a particular method. This design easily evolves into per action controls across the board. 
+*  👎👎 Breaking API Change for Budgets at least, and if we decide to model DisruptAfter we also would have to break those apis. It might make sense to break the budgets now before they have garnered large adoption as it becomes harder to make the change as time goes on.
+* 👎 Doesn't allow for easy defaulting of `All` disruption actions
+* 👎 Adds complexity to the use of budgets, before its high level on the disruption controls but with this design approach its nested inside another field.
+
+### API Design Conclusion: Preferred Design
+If the goal is to provide a simple, backward-compatible solution with immediate applicability, Approach A is more suitable. It provides a straightforward way to manage disruptions without overhauling the existing system. Breaking API changes in Approach B are likely too disruptive to customers. 
+
+## Clarifying Behavioral Questions 
+### Q: How should Karpenter handle the default or undefined action case? 
+The current design involves specifying a specific number of disruptable nodes per action, which can complicate the disruption lifecycle. For example, if there's a 10-node budget for "Drift" and a separate 10-node budget for "Consolidation", but a 15-node budget for "All" determining which nodes will get disrupted becomes unclear. Would it be 10 nodes for "Drift" and 5 nodes for "Consolidation"?
+
+We could consider treating an undefined action as a budget for all disruption actions except for those with explicitly defined budgets. In this scenario, if a user specifies a disruption budget like this:
+
+```yaml
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    budgets:
+    - schedule: "* * * * *"
+      action: Consolidation
+      nodes: 10
+    - schedule: "* * * * *"
+      action: Drift
+      nodes: 10
+    # For all other actions , only allow 5 nodes to be disrupted at a time
+    - nodes: 5
+      schedule: "* * * * *"
+```
+
+It means that "Consolidation" and "Drift" actions have specific budgets of 10 nodes each, while all other actions (e.g., expiration and emptiness) share a common budget of 5 nodes. This approach simplifies the configuration but has one limitation: it may not allow the execution of other disruption actions if a specific action exhausts the budget. This is a problem with the existing design for disruption budgets.
```

Comment:
Will think about this case some more, and provide more thoughts. Haven't fully made up my mind here.

Reply:
> I personally think this is a concerning degree of "required" knowledge for what budgets override what. Although, maybe its fine. I just think that the safest approach is to treat "All" as applying to all, so you have the ability to have the budgets conservative and rely on the most restrictive setting.

Maybe All should be called default instead. If you don't have a method specifed its waht is used. and it will just be rpresetned as an empty string or unspecified method name in the budgets.

Reply:
internally in the DisruptionBudgetMapping we can use a constant default = ""

Reply:
I think default is more clear than "all"

Reply:
decided on default